#### Import packages ####
library(openxlsx)
library(tidyverse)
library(edgeR)
library(limma)
library(vegan)
library(moments)
library(ggplot2)
library(rtracklayer)
library(mclust)
library(ggpubr)
library(sva)
library(ggstream)
library(reshape2)
library(ggsci)


#### Functions ####
# number of zero-expressing gene for each sample
calculating_ZeroRPKMGeneCount <- function(rpkm_dat) {
  t_rpkm_dat <- as.data.frame(t(rpkm_dat))
  t_rpkm_dat$ZeroGeneCount <- apply(t_rpkm_dat,1,function(x){length(x[x==0])})
  zero_dat <- data.frame(Run = rownames(t_rpkm_dat), ZeroGeneCount = t_rpkm_dat$ZeroGeneCount)
  return(zero_dat)
}

# min/max, sd, mean, median, q25/75/IQR, SI, kurtosis, skewness, zero-expression
calculating_ExprAnnotation <- function(expr, # RPKM
                                       log = T
) {
  # whether log expression data
  if (log) {expr <- log2(expr+1)}
  
  # min, max
  expr.min <- apply(expr, 1, min)
  expr.max <- apply(expr, 1, max)
  
  # SD, mean, median of log2(RPKM+1)
  expr.sd.logRPKM <- apply(expr, 1, sd)
  expr.mean.logRPKM <- apply(expr, 1, mean)
  expr.median.logRPKM <- apply(expr, 1, median)
  
  # median and IQR(log2(RPKM+1))
  expr.q25.logRPKM <- apply(expr, 1, function(x) {quantile(x,0.25)})
  expr.q75.logRPKM <- apply(expr, 1, function(x) {quantile(x,0.75)})
  expr.IQR.logRPKM <- expr.q75.logRPKM - expr.q25.logRPKM
  
  # Shannon-Wiener index (using log2(RPKM+1))
  expr.shannon <- vegan::diversity(expr, index = "shannon")
  
  # kurtosis, skewness
  expr.kurtosis <- apply(expr, 1, moments::kurtosis)
  expr.skewness <- apply(expr, 1, moments::skewness)
  
  # Zero-expression proportion
  expr.zero.per <- apply(expr, 1, function(x) {length(x[x==0])})
  expr.zero.per <- expr.zero.per / ncol(expr)
  
  # MERGE
  gene.ExprAnnotation <- data.frame(gene = rownames(expr),
                                    MinExpr = expr.min, MaxExpr = expr.max,
                                    SD_logRPKM = expr.sd.logRPKM, 
                                    Mean_logRPKM = expr.mean.logRPKM,
                                    Median_logRPKM = expr.median.logRPKM,
                                    Q.25_logRPKM = expr.q25.logRPKM,
                                    Q.75_logRPKM = expr.q75.logRPKM,
                                    IQR_logRPKM = expr.IQR.logRPKM,
                                    ShannonIndex = expr.shannon,
                                    Kurtosis = expr.kurtosis, 
                                    Skewness = expr.skewness,
                                    ZeroExpr_proportion = expr.zero.per)
  
  return(gene.ExprAnnotation)
}

# TP
calculating_TP <- function(ExprAnnotation_data, #  generated by FUNCTION "calculating_ExprAnnotation"
                           SI_higher_than = 0, # select genes with Shannon index > `SI_higher_than`
                           ZeroExprPercent_lower_than = 1, # select genes with Zero-expressed proportion < `ZeroExprPercent_lower_than`
                           loess_span = 0.7,
                           loess_degree = 1
) {
  ExprAnnotation <- ExprAnnotation_data[ExprAnnotation_data$ShannonIndex > SI_higher_than & 
                                          ExprAnnotation_data$ZeroExpr_proportion < ZeroExprPercent_lower_than,]
  loess.model <- loess(SD_logRPKM ~ Mean_logRPKM, data = ExprAnnotation, span = loess_span, degree = loess_degree)
  ExprAnnotation$residual.SD_logRPKM <- loess.model$residuals
  ExprAnnotation$LOESS.fit <- loess.model$fitted
  ExprAnnotation$TP <- ExprAnnotation$residual.SD_logRPKM + mean(ExprAnnotation$LOESS.fit)
  
  return(ExprAnnotation)
}

# fig of SI/zero-expr and sd
SI_ZeroExpr_SD <- function(ExprAnnotation_data) { #  generated by FUNCTION "calculating_ExprAnnotation"
  p <- ggarrange(ggplot(ExprAnnotation_data, aes(x = ZeroExpr_proportion)) + geom_histogram(bins = 100),
                 ggplot(ExprAnnotation_data, aes(x = ShannonIndex)) + geom_histogram(bins = 100),
                 ggplot(ExprAnnotation_data, aes(x = ZeroExpr_proportion, y = ShannonIndex)) + geom_point(size = 0.1, color = "red"),
                 ggplot(ExprAnnotation_data, aes(x = ZeroExpr_proportion, y = SD_logRPKM)) + geom_point(size = 0.1, color = "red"),
                 ggplot(ExprAnnotation_data, aes(x = ZeroExpr_proportion, y = Mean_logRPKM)) + geom_point(size = 0.1, color = "red"),
                 ggplot(ExprAnnotation_data, aes(x = ZeroExpr_proportion, y = SD_logRPKM/Mean_logRPKM)) + geom_point(size = 0.1, color = "red"),
                 ggplot(ExprAnnotation_data, aes(x = ShannonIndex, y = SD_logRPKM)) + geom_point(size = 0.1, color = "red"),
                 ggplot(ExprAnnotation_data, aes(x = ShannonIndex, y = Mean_logRPKM)) + geom_point(size = 0.1, color = "red"),
                 ggplot(ExprAnnotation_data, aes(x = ShannonIndex, y = SD_logRPKM/Mean_logRPKM)) + geom_point(size = 0.1, color = "red"),
                 nrow = 3,ncol = 3)
  return(p)
}


#### Input data ####
unfiltered_rpkm_mtb <- read.csv("./data/unfiltered tmm rpkm Mtb.csv", row.names = 1)
unfiltered_rpkm_msm <- read.csv("./data/unfiltered tmm rpkm Msm.csv", row.names = 1)
unfiltered_rpkm_mab <- read.csv("./data/unfiltered tmm rpkm Mab.csv", row.names = 1)


#### Remove samples with higher proportion zero-expressing genes ####
# calculate frequency of zero-expressing genes in each sample
ZeroGeneCount_Mtb <- calculating_ZeroRPKMGeneCount(unfiltered_rpkm_mtb)
ZeroGeneCount_Msmeg <- calculating_ZeroRPKMGeneCount(unfiltered_rpkm_msm)
ZeroGeneCount_Mab <- calculating_ZeroRPKMGeneCount(unfiltered_rpkm_mab)

# remove samples with more than 4% zero-expressing genes
Threshold_ZeroGene_per <- 0.04 

final_sample_Mtb <- ZeroGeneCount_Mtb$Run[ZeroGeneCount_Mtb$ZeroGeneCount < Threshold_ZeroGene_per*4008] # Mtb has 4,008 genes
final_sample_Msmeg <- ZeroGeneCount_Msmeg$Run[ZeroGeneCount_Msmeg$ZeroGeneCount < Threshold_ZeroGene_per*6884] # Msm has 6,884 genes
final_sample_Mab <- ZeroGeneCount_Mab$Run[ZeroGeneCount_Mab$ZeroGeneCount < Threshold_ZeroGene_per*4942] # Mab has 4942 genes

samplefiltered_rpkm_mtb <- unfiltered_rpkm_mtb[,colnames(unfiltered_rpkm_mtb) %in% final_sample_Mtb]
samplefiltered_rpkm_msm <- unfiltered_rpkm_msm[,colnames(unfiltered_rpkm_msm) %in% final_sample_Msmeg]
samplefiltered_rpkm_mab <- unfiltered_rpkm_mab[,colnames(unfiltered_rpkm_mab) %in% final_sample_Mab]


#### Measure gene expression patterns ####
ExprAnno_Mtb <- calculating_ExprAnnotation(samplefiltered_rpkm_mtb)
ExprAnno_Msmeg <- calculating_ExprAnnotation(samplefiltered_rpkm_msm)
ExprAnno_Mab <- calculating_ExprAnnotation(samplefiltered_rpkm_mab)

# plot
SI_ZeroExpr_SD(ExprAnno_Mtb)
SI_ZeroExpr_SD(ExprAnno_Msmeg)
SI_ZeroExpr_SD(ExprAnno_Mab)

dev.off()


#### Filtrate genes and calculate TP ####
# set threshold for filtering genes based on Shannon index and zero-expressing proportion (plots above)
SI_mtb <- 6.5
ZeroExpPer_mtb <- 0.1

SI_msm <- 4
ZeroExpPer_msm <- 0.1

SI_mab <- 4
ZeroExpPer_mab <- 0.1

# calculate TP after removing genes with low Shannon index and high zero-expressing proportion
TP_Mtb <- calculating_TP(ExprAnnotation_data = ExprAnno_Mtb,
                         SI_higher_than = SI_mtb, ZeroExprPercent_lower_than = ZeroExpPer_mtb)

TP_Msmeg <- calculating_TP(ExprAnnotation_data = ExprAnno_Msmeg, 
                           SI_higher_than = SI_msm, ZeroExprPercent_lower_than = ZeroExpPer_msm)

TP_Mab <- calculating_TP(ExprAnnotation_data = ExprAnno_Mab,
                         SI_higher_than = SI_mab, ZeroExprPercent_lower_than = ZeroExpPer_mab)

write.csv(TP_Mtb, "./data/TP Mtb.csv")
write.csv(TP_Msmeg, "./data/TP Msmeg.csv")
write.csv(TP_Mab, "./data/TP Mab.csv")


#### Filtered transcriptomic profiles ####
filtered_rpkm_mtb <- samplefiltered_rpkm_mtb[rownames(samplefiltered_rpkm_mtb) %in% TP_Mtb$gene,]
filtered_rpkm_msm <- samplefiltered_rpkm_msm[rownames(samplefiltered_rpkm_msm) %in% TP_Msmeg$gene,]
filtered_rpkm_mab <- samplefiltered_rpkm_mab[rownames(samplefiltered_rpkm_mab) %in% TP_Mab$gene,]

write.csv(filtered_rpkm_mtb,"./data/RPKM Mtb.csv")
write.csv(filtered_rpkm_msm,"./data/RPKM Msm.csv")
write.csv(filtered_rpkm_mab,"./data/RPKM Mab.csv")

